src/QsCompiler/QirGeneration/Context.cs

```diff
@@ -1131,33 +1131,49 @@ Value BuildLoadForArg(QsArgumentTuple arg, Value value)
                 return args;
             }
 
-            void PopulateResultTuple(ResolvedType resultType, Value resultValue, Value resultTuplePointer, int item)
+            // result value contains the return value, and output tuple is the tuple where that value should be stored
+            void PopulateResultTuple(ResolvedType resultType, Value resultValue, Value outputTuple)
             {
                 var resultTupleType = this.LlvmStructTypeFromQsharpType(resultType);
+                Value GetOutputItem(int item) =>
+                    this.CurrentBuilder.GetElementPtr(
+                        resultTupleType,
+                        this.CurrentBuilder.BitCast(outputTuple, resultTupleType.CreatePointerType()),
+                        new[] { this.Context.CreateConstant(0L), this.Context.CreateConstant(item) });
+
                 if (resultType.Resolution is QsResolvedTypeKind.TupleType tupleType)
                 {
-                    // Here we'll step through and recurse
-                    var itemCount = tupleType.Item.Length;
                     // Start with 1 because the 0 element of the LLVM structures is the tuple header
-                    for (int i = 1; i <= itemCount; i++)
+                    for (int i = 1; i <= tupleType.Item.Length; i++)
                     {
-                        // TODO: complete implementation
+                        var resItem = this.CurrentBuilder.GetElementPtr(
+                             resultTupleType,
+                             resultValue,
+                             new[] { this.Context.CreateConstant(0L), this.Context.CreateConstant(i) });
+                        var itemOutputPointer = GetOutputItem(i);
+
+                        // if the subitem is a tuple we are good
+                        // if it is not then we need to get an output tuple to populate for the recursion to work... -> change recursion?
+                        if (tupleType.Item[i].Resolution.IsTupleType)
+                        {
+                            PopulateResultTuple(tupleType.Item[i], resItem, itemOutputPointer);
```

Comment:
Don't you need to create a new tuple before this, based on the item type, and store the pointer to the new tuple in itemOutputPointer?
And resItem is a pointer to the pointer to the source tuple, so you need to Load from it to get a pointer to the source tuple.
When the item is a tuple, which is to say a pointer to a struct, itemOutputPointer will be a pointer to a pointer to the new copy, not a pointer to the new copy.

Reply:
Ah, yes, I did not read the code sufficiently to realize that there is no space allocated for the output tuple yet.

Reply:
Actually, since the tuples already exists I think it is sufficient to fill in the pointer to the existing tuple (the on in resultValue) into the right location of the output tuple, much like non-tuple items. I hence think the case distinction or any kind of recursion is not actually necessary, unless I am somehow missing that the tuple constructed as part of building the resultValue gets deallocated somewhere? @alan-geller Could you take another look?

Reply:
Yes, that should work, you can just copy the tuple pointer. The only subtlety is that you need to make sure you add a new reference to the tuple; you can use the scope manager to make sure this happens.

Reply:
It is not currently doing that for other references like callables or arrays either. I think we need to take a closer look at the reference counting in any case.

Reply:
After taking a closer look: we do not unreference the return value, and I hence believe there is no need to add an additional reference here.